modules/3d/src/rendering.cpp

```diff
@@ -0,0 +1,333 @@
+                    {
+                        color = colors[0];
+                    }
+                    else // ShadingType::Shaded
+                    {
+                        Vec3f interp;
+                        for (int i = 0; i < 3; i++)
+                        {
+                            interp[i] = f[i] * zInter * invz[i];
+                        }
+
+                        for (int i = 0; i < 3; i++)
+                        {
+                            for (int j = 0; j < 3; j++)
+                            {
+                                color[i] += interp[j] * colors[j][i];
+                            }
+                        }
+                    }
+                    colorBuf.at<Vec3f>(y, x) = color;
+                }
+            }
+        }
+    }
+}
+
+
+CV_EXPORTS  void triangleRasterize(InputArray _vertices, InputArray _indices, InputArray _colors,
+                                   InputArray cameraPose, float fovyRadians, float zNear, float zFar,
+                                   int width, int height, RasterizeSettings settings,
+                                   OutputArray _depthBuffer, OutputArray _colorBuffer)
+{
+    bool needDepth = _depthBuffer.needed();
+    bool needColor = _colorBuffer.needed();
+
+    if (!needDepth && !needColor)
+    {
+        CV_Error(Error::StsBadArg, "No depth nor color output image provided");
+    }
+
+    CV_Assert(cameraPose.type() == CV_32F);
+    CV_Assert((cameraPose.size() == Size {4, 3}) || (cameraPose.size() == Size {4, 4}));
+
+    CV_Assert((fovyRadians > 0) && (fovyRadians < CV_PI));
+    CV_Assert(zNear > 0);
+    CV_Assert(zFar > zNear);
+    CV_Assert((width > 0) && (height > 0));
+
+    Mat cpMat = cameraPose.getMat();
+    Matx44f camPoseMat = Matx44f::eye();
+    for (int i = 0; i < 3; i++)
+    {
+        for (int j = 0; j < 4; j++)
+        {
+            camPoseMat(i, j) = cpMat.at<float>(i, j);
+        }
+    }
+
+    bool hasIdx    = !_indices.empty();
+    bool hasColors = !_colors.empty();
+
+    Mat vertices, colors, triangles;
+    int nVerts = 0, nColors = 0, nTriangles = 0;
+
+    if (hasIdx)
+    {
+        CV_CheckFalse(_vertices.empty(), "No vertices provided along with indices array");
+
+        int vertexType = _vertices.type();
+        CV_Assert(vertexType == CV_32FC1 || vertexType == CV_32FC3);
+        vertices = _vertices.getMat();
+        if ((_vertices.channels() == 1) && (_vertices.rows() == 3))
+        {
+            vertices = vertices.t();
+        }
+        vertices = vertices.reshape(3, 1).t();
+        nVerts = vertices.total();
+
+        int indexType = _indices.type();
+        CV_Assert(indexType == CV_32SC1 || indexType == CV_32SC3);
+        triangles = _indices.getMat();
+        if ((_indices.channels() == 1) && (_indices.rows() == 3))
+        {
+            triangles = triangles.t();
+        }
+        triangles = triangles.reshape(3, 1).t();
+        nTriangles = triangles.total();
```

Comment:
`triangles.t();` happens 2 times in some cases.

Reply:
That second transposition is performed on 1xN matrix so it's almost free in terms of performance